#35: XSLDocument.SPLIT javadoc makes it clear that it strips spaces of w...

pinaf opened the pull request:
#35

dmarkov commented:
Thanks for your pull request, let me find someone who can review it

dmarkov commented:
@krzyk it's yours, please review

pinaf commented:
@krzyk @yegor256 @dmarkov 

Travis
`
[INFO] Caused by: org.eclipse.aether.resolution.DependencyResolutionException: Could not transfer artifact net.sourceforge.saxon:saxon:jar:9.1.0.8 from/to central (http://repo.maven.apache.org/maven2): GET request of: net/sourceforge/saxon/saxon/9.1.0.8/saxon-9.1.0.8.jar from central failed

[INFO] Caused by: java.net.SocketException: Connection reset
`

Can we run the CI tests again?

pinaf commented:
@yegor256 @dmarkov  Looks like the existing test for STRIP doesn't work in AppVeyor, probably because of the `\n` character in a Windows environment. Is this something I should report as a bug?

krzyk commented:
it build with oraclejdk7 and openjdk6 in travis so should also work in other cases (minus the network issues that sometimes happen), let see how rultor will handle it

krzyk commented:
@rultor merge pls

rultor commented:
> @rultor merge pls

@krzyk Thanks for your request. @yegor256 Please confirm this.

pinaf commented:
@krzyk yep, I agreee that it should pass Travis no problem. However, the issue with AppVeyor remains and seems to be related to line breaks in a Windows environment.

yegor256 commented on src/main/java/com/jcabi/xml/XSLDocument.java:
```diff
@@ -70,7 +70,8 @@
     /**
-     * Strip spaces.
+     * Strips spaces of whitespace-only text nodes. This will NOT remove
+     * line-breaks between Element nodes.
```
this looks rather short. in a few months someone will ask the same question - why not remove line-breaks? and will submit a bug. I think we should put a longer explanation, like you did in the ticket. To avoid confusion.

yegor256 commented:
@pinaf I think we need a longer and more detailed explanation, see my comment

pinaf commented:
@yegor256 I decided to spend more time on this and there is a way to solve it that makes sense. May I suggest we don't automatically add indentation to the internal immutable String representation (the `xml` field) of a `XMLDocument`? That is the only indentation that doesn't get removed by the STRIP XSL transformation - getting rid of that means I can implement #35 . Also, it makes sense that the internal representation is as compact as possible and that an indented String is used just for display purposes.

I traced back the commit that created this indentation and it was done when `XMLDocument` was made immutable in issue #12 but the code there was taken from a previous implementation of the `toString` method, which makes sense because that is used to display a pretty XML output.

yegor256 commented:
@pinaf I see what you mean and I'm totally in favor of the changes you're suggesting. Please, try and let's see whether the build will stay clean

pinaf commented:
@yegor256 needed to make one small change to test `XMLDocumentTest#comparesToAnotherDocument` which erroneously considered two different XMLs equal (because of the pretty indent that was present when instantiating a `XMLDocument`).

build success with Pqulice. now let me implement the original functionality.

pinaf commented:
@yegor256 After adding the new `compact` method PMD complains about too many methods.
```
[ERROR] /src/main/java/com/jcabi/xml/StrictXML.java[64-245]: This class has too many methods, consider refactoring it. (TooManyMethods)
```

Suggestions?

yegor256 commented:
@pinaf suppress it, as explained here http://www.qulice.com/quality.html

pinaf commented:
@krzyk @yegor256 please review. desired functionality is provided by the `XML#compact` method.

yegor256 commented:
@krzyk take a look again pls

pinaf commented:
looks like appveyor is stuck.

krzyk commented on src/main/java/com/jcabi/xml/XSLDocument.java:
```diff
@@ -71,6 +71,7 @@
 
     /**
      * Strip spaces.
+     * line-breaks between Element nodes.
```
there is something missing in the description :)

pinaf replied:
@krzyk typo :(

krzyk commented on src/test/java/com/jcabi/xml/XMLDocumentTest.java:
```diff
@@ -349,9 +349,10 @@ public void buildsDomNode() throws Exception {
         MatcherAssert.assertThat(
-            new XMLDocument("<hi>\n<dude>  </dude></hi>"),
-            Matchers.equalTo(new XMLDocument("<hi><dude>  </dude></hi>"))
+            new XMLDocument(xml),
```
this looks strange, we are comparing two instances of `XMLDocument` that are created using the same parameter, they will be always equal.
Can't we just leave the previus parameters here? (with new lines)

pinaf replied:
Not really. Now that the String `xml` doesn't get automatic indentation, the document with the `\n` is NOT equal to the document without the `\n`. In order for equals to be true, we have to compare instances of XMLDocument constructed with the same String.

krzyk replied:
There is no new line in this test, here is what you do:
``
 MatcherAssert.assertThat(new XMLDocument(xml), Matchers.equalTo(new XMLDocument(xml)));
``

pinaf replied:
Without indentation in the final `XMLDocument(xml)));` part?

krzyk replied:
It is confusing to have unrelated asserts in one test method, please split those into separate tests.

pinaf replied:
@krzyk this test `comparesToAnotherDocument` already had two comparisons, one for equals and one for not equals. All I did was to fix the test because of the `\n`. You think it doesn't make sense to teste equals and not equals in the same test?

krzyk replied:
Right, forgot that it is already existing method, lets leave it.

krzyk commented on src/test/java/com/jcabi/xml/XMLDocumentTest.java:
```diff
@@ -374,4 +375,32 @@ public void preservesXmlNamespaces() throws Exception {
+            Matchers.not(
+                Matchers.equalTo(
+                    XML.class.cast(
+                        new XMLDocument("<a><b/>xyz<c> text </c></a>")
```
I don't get it why we compare it to a document that has text inserted into it

pinaf replied:
You mean the string ` text ` inside of element `<c>`? Just to make it clear that the text content will keep the spaces it has (because the node contains non-whitespace characters), while the text nodes with just whitespaces such as ` \n\n\t ` will get removed. A question of clarity.

krzyk replied:
No, I mean `xyz` which appeared between `<b/>` and `<c>`

pinaf replied:
Just a way to get a different document for the not equals test.

krzyk replied:
Similarly here, the first assert is not related to the test, let's move it to another method that will test inequality.

pinaf replied:
@krzyk trying to make sure `compact()` does what it is supposed to by making sure it is equal to what it is supposed to be and not equal to something different. Don't you think that is related?

krzyk replied:
It would make more sense to make sure it is not equal to the uncompacted XML.

pinaf replied:
agreed.

krzyk commented:
@pinaf Very good find, I have  some comments above

pinaf commented:
@krzyk thanks. fixed the typo in the first comment and responded to your other comments. Can you take a look again?

pinaf commented:
@krzyk fixed. please look at it again.

krzyk commented:
@rultor merge pls

rultor commented:
> @rultor merge pls

@krzyk Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/main/java/com/jcabi/xml/StrictXML.java:
```diff
@@ -139,6 +140,9 @@ public Node node() {
     }
 
+    @Override
+    public XML compact() { return this.origin.compact(); }
```
hm.. we don't use this syntax. put this `return` on the next line pls

yegor256 commented on src/main/java/com/jcabi/xml/XMLDocument.java:
```diff
@@ -284,6 +284,10 @@ public String toString() {
 
     @Override
+    @NotNull(message = "compact is never NULL")
+    public XML compact() { return XSLDocument.STRIP.transform(this); }
```
same here

pinaf commented:
@yegor256 fixed.

yegor256 commented:
@pinaf brr... first, why do we need this `compact()` method if we have `SPLIT`? that's absolutely redundant. anyone who wants to compact, should use `SPLIT` and that's it. second, the entire library will be affected since you remove indentation. all XML documents will NOT be indented from now on. it's a very serious change. I don't think it's correct to do it like this

pinaf commented:
@yegor256 the `compact()` method was requested in the issue. `STRIP` is simply the current implementation as you suggested at first, but `compact()` might involve other transformations in the future. The code is also clearer IMHO.

On the issue of the XML documents not being indented, I agree it is a serious change and that is why I asked you first. Your response was

>@pinaf I see what you mean and I'm totally in favor of the changes you're suggesting. Please, try and let's see whether the build will stay clean

This change has the upside of making the internal representation more lightweight and an indented String would be used just for display purposes.

yegor256 commented:
@pinaf nope, we can't change the way documents looks from `toString()`. they have to look nice/indented. that's the purpose of the library. the only option I see is to document `STRIP` properly, explaining that we can't compress tighter than an indented document. If a user needs tighter compression, he is on his own. He has to use Java XML transformation API.

pinaf commented:
@yegor256 we could also make `toString()` return a pretty indented String. your call. change javadoc or toString?

pinaf commented:
@yegor256 also, if we go with changing just the javadoc for `STRIP` should we keep the compact method in interface `XML`?

yegor256 commented:
@pinaf but we can't return pretty-print from `toString()` and encapsulate compressed XML, you explained that before. We pretty-print during encapsulation. Then, we have no way back. Or I'm missing something? `compact()` method makes no sense in its current form.

pinaf commented:
@yegor256 I think we could. What we can't do is encapsulate with `indent=yes`, like we do now, and then try to remove the indentation with the STRIP transform - it doesn't remove the spaces. We could, in theory, encapsulate the String with `indent=no` and also already STRIPPED and, in `toString()`, transform it to a pretty representation with a simple XSL identity transformation + `indent=yes`.

pinaf commented:
@yegor256 anyway, I added the Javadoc info as you requested and removed the `compact()` method. Please review.

Should I add puzzles to implement the functionality above (compacted internal representation, prettified `toString()`)?

yegor256 commented:
@rultor merge it

rultor commented:
> @rultor merge it

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/1587-70273919)

yegor256 commented:
@pinaf I think that's enough for this task. We explained the problem and analyzed it. No puzzles needed, we won't implement this functionality. This will make the library too complex.

rultor commented:
> @rultor merge it

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/1587-70273919) (took me 7min)

dmarkov commented:
@krzyk thank you, added **16 mins** to your acc, payment num is `50519754`... +16 added to your rating, current score is: [+2051](http://www.netbout.com/b/35035?open=rating)

dmarkov commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@dmarkov OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/1587-70641062)

rultor commented:
> @rultor please deploy

@dmarkov Done! FYI, the full log is [here](http://www.rultor.com/t/1587-70641062) (took me 7min)
